### PR TITLE
Move slerp and scaling of error to 'cost_helpers'.

### PR DESCRIPTION
**Pull request by pifon2a**

### CODE REVIEW

File: cartographer/mapping/pose_graph/cost_helpers_impl.h [hunk 1]
```diff
@@ -69,6 +79,43 @@ static std::array<T, 6> ComputeUnscaledError3d(
...
+  const T cos_theta = start[0] * end[0] + start[1] * end[1] +
+                      start[2] * end[2] + start[3] * end[3];
+  // If numerical error brings 'cos_theta' outside of [-1., 1.] interval, then
+  // the quaternions are likely to be collinear.
+  if (cos_theta >= T(1.0) || cos_theta <= T(-1.0)) {
+    return {{end[0], end[1], end[2], end[3]}};
+  }
+  const T theta = acos(abs(cos_theta));
+  const T sin_theta = sin(theta);
+  const T prev_scale = sin((T(1.0) - factor) * theta) / sin_theta;
```
Comment (gaschler) [inline cartographer/mapping/pose_graph/cost_helpers_impl.h]: drive-by comment: is sin_theta always non-zero?

File: cartographer/mapping/pose_graph/cost_helpers_impl.h [hunk 2]
```diff
@@ -92,24 +92,27 @@ std::array<T, 6> ScaleError(std::array<T, 6> error, T translation_weight,
...
-  // If numerical error brings 'cos_theta' outside of [-1., 1.] interval, then
-  // the quaternions are likely to be collinear.
-  if (cos_theta >= T(1.0) || cos_theta <= T(-1.0)) {
-    return {{end[0], end[1], end[2], end[3]}};
+  const T abs_cos_theta = abs(cos_theta);
+  // If numerical error brings 'cos_theta' outside [-1 + epsilon, 1 - epsilon]
+  // interval, then the quaternions are likely to be collinear.
+  T prev_scale = T(1.) - factor;
+  T next_scale = factor;
+  if (abs_cos_theta < T(1. - 1e-5)) {
```
Comment (gaschler) [inline cartographer/mapping/pose_graph/cost_helpers_impl.h]: Looks good.

Comment (pifon2a) [reply cartographer/mapping/pose_graph/cost_helpers_impl.h]: Thanks for noticing this division by 0.

Comment (gaschler) [reply cartographer/mapping/pose_graph/cost_helpers_impl.h]: Interesting that Eigen does not handle this case specially https://bitbucket.org/eigen/eigen/src/2355b229ea4c2876f490e726526cdcd8a63c7f54/Eigen/src/Geometry/Quaternion.h?at=default&fileviewer=file-view-default#Quaternion.h-748
